src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -2138,6 +2153,26 @@ void CheckClosed()
         }
 
         #endregion
+
+        #region Misc
+
+        /// <summary>
+        /// Unbinds reader from the connector.
+        /// Should be called before the connector is returned to the pool.
+        /// </summary>
+        internal void UnbindIfNecessary()
+        {
+            // We're closing the connection, but reader is not yet disposed, the connector isn't broken and pooling is on
+            // We have to unbind the reader from the connector, otherwise there could be a concurency issues
+            // See #3126 and #3290
+            if (!_isDisposed && !Connector.IsBroken && Connector.Settings.Pooling)
```

Comment:
I wonder if the Pooling check is a good idea. If we don't do it, then the non-pooling case recycles in this edge although it didn't need to (very negligible). If we do this check, we do it all the time for all pooled usage - this probably doesn't make sense, no?